### PR TITLE
ARROW-11377: [C++][CI] Add Thread Sanitizer nightly build

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit test-ubuntu-20.04-cpp-thread-sanitizer

**Comment by pitrou** [comment]
cc @bkietz

File: cpp/build-support/run-test.sh [hunk 1]
```diff
@@ -65,14 +65,9 @@ function setup_sanitizers() {
 
   # Configure TSAN (ignored if this isn't a TSAN build).
   #
-  # Deadlock detection (new in clang 3.5) is disabled because:
-  # 1. The clang 3.5 deadlock detector crashes in some unit tests. It
-  #    needs compiler-rt commits c4c3dfd, 9a8efe3, and possibly others.
-  # 2. Many unit tests report lock-order-inversion warnings; they should be
-  #    fixed before reenabling the detector.
-  TSAN_OPTIONS="$TSAN_OPTIONS detect_deadlocks=0"
-  TSAN_OPTIONS="$TSAN_OPTIONS suppressions=$ROOT/build-support/tsan-suppressions.txt"
```
Comment (bkietz) [inline cpp/build-support/run-test.sh:74]: Is it correct to remove these suppressions? If so, why not also remove tsan-suppressions.txt?

**Review by bkietz — COMMENTED** [review]
LGTM (and thanks for adding this), I just have a concern re: suppressions

File: cpp/build-support/run-test.sh (hunk 1 above)
Comment (pitrou) [reply cpp/build-support/run-test.sh:74]: Hmm, no, the removal is a mistake. Will re-add them (they're for CUDA).

**Comment by pitrou** [comment]
@github-actions crossbow submit test-ubuntu-20.04-cpp-thread-sanitizer

**Comment by pitrou** [comment]
The Crossbow test failure is expected.

**Pull request closed without merging**